### PR TITLE
Fix wrong typing for tolerance in reindex

### PR DESCRIPTION
**Pull request by antscloud**
In the `xarray.core.dataset.py `module, more particulary in the reindex method the tolerance argument is set to be a `Number`

https://github.com/pydata/xarray/blob/f08672847abec18f46df75e2f620646d27fa41a2/xarray/core/dataset.py#L2743

But the `_reindex`  function call  `reindex_variable` function. In the `reindex_variable`  the type for tolerance is `Any`. This function ends to call `get_indexer_nd` which call the pandas function `get_indexer` :

https://github.com/pydata/xarray/blob/f08672847abec18f46df75e2f620646d27fa41a2/xarray/core/indexes.py#L137

In pandas the type of tolerance according to the docs can be a scalar or a list-like object

- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: xarray/tests/test_dataset.py [hunk 1]
```diff
@@ -1894,6 +1894,10 @@ def test_reindex_method(self):
         expected = Dataset({"x": ("y", 3 * [np.nan]), "y": y})
         assert_identical(expected, actual)
 
+        actual = ds.reindex(y=y, method="backfill", tolerance=[0.1, 0.5, 0.1])
+        expected = Dataset({"x": ("y", [np.nan, 20, np.nan]), "y": y})
+        assert_identical(expected, actual)
+
```
Comment (mathause) [inline xarray/tests/test_dataset.py:1900]: You could add a test where the tolerance makes a difference:

```suggestion

actual = ds.reindex(y=[0.1, 0.1, 1], tolerance=[0, 0.1, 0], method="nearest")
expected = Dataset({"x": ("y", [np.nan, 10, 20]), "y": [0.1, 0.1, 1]})
assert_identical(expected, actual)
```

**Review by mathause — APPROVED** [review]
Thanks for your patience! This looks good. One minor suggestion.

**Comment by max-sixty** [comment]
Thanks a lot!

On the relevant test, could we please add `-> None` to the `def`? That will cause mypy to use the test to check the types.

**Comment by antscloud** [comment]
Thank you too for your patience 
@max-sixty I don't understand where to put the `-> None`
In the test function ?
https://github.com/pydata/xarray/blob/118a00b1233db34fe90072ed2e21501337b04b35/xarray/tests/test_dataset.py#L1886

How does mypy can test these types ?

**Comment by Illviljan** [comment]
```python
def test_reindex_method(self) -> None:
```
Should do the trick.

**Comment by Illviljan** [comment]
Some typing errors now pop up now, now that we're testing it. Probably easier to use int and float instead?
```
xarray/tests/test_dataarray.py:1534: error: Argument "tolerance" to "reindex" of "DataArray" has incompatible type "float"; expected "Union[Number, Iterable[Number], None]"  [arg-type]
xarray/tests/test_dataarray.py:1538: error: List item 0 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataarray.py:1538: error: List item 1 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataarray.py:1538: error: List item 2 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataset.py:1893: error: Argument "tolerance" to "reindex" of "Dataset" has incompatible type "float"; expected "Union[Number, Iterable[Number], None]"  [arg-type]
xarray/tests/test_dataset.py:1897: error: List item 0 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataset.py:1897: error: List item 1 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataset.py:1897: error: List item 2 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataset.py:1901: error: List item 0 has incompatible type "int"; expected "Number"  [list-item]
xarray/tests/test_dataset.py:1901: error: List item 1 has incompatible type "float"; expected "Number"  [list-item]
xarray/tests/test_dataset.py:1901: error: List item 2 has incompatible type "int"; expected "Number"  [list-item]
```

**Comment by Illviljan** [comment]
`Number` is apparently a no-go for `int` and `float` when type checking.

For a future PR, but I don't think `.sel` is being type-checked properly in tests either as it uses Number as well:
https://github.com/pydata/xarray/blob/4c865d607e6e03605b7050d9fb6991e86346bf08/xarray/core/dataset.py#L2446

**Review by Illviljan — APPROVED** [review]

**Comment by max-sixty** [comment]
Thank you @antscloud !

**Pull request merged**